onnx/defs/generator/defs.cc

```diff
@@ -896,4 +896,76 @@ ONNX_OPERATOR_SET_SCHEMA(
           }
         }));
 
+static const char* Bernoulli_ver15_doc = R"DOC(
+    Draws binary random numbers (0 or 1) from a Bernoulli distribution. The input tensor should be
+    a tensor containing probabilities to be used for drawing the binary random number.
+    )DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Bernoulli,
+    15,
+    OpSchema()
+        .SetDoc(Bernoulli_ver15_doc)
+        .Attr(
+            "seed",
+            "(Optional) Seed to the random generator, if not specified we will auto generate one.",
+            AttributeProto::FLOAT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "dtype",
+            "The data type for the elements of the output tensor. if not specified, we will use "
+            "the data type of the input tensor.",
+            AttributeProto::INT,
+            OPTIONAL_VALUE)
+        .Input(
+            0,
+            "input",
+            "All values in input have to be in the range:[0, 1].",
+            "T1")
+        .Output(
+            0,
+            "output",
+            "The returned output tensor only has values 0 or 1, same shape as input tensor.",
+            "T2")
+        .TypeConstraint(
+            "T1",
+            {"tensor(float16)",
+              "tensor(float)",
+              "tensor(double)",
+              "tensor(bfloat16)"},
+              "Constrain input types to float tensors.")
+        .TypeConstraint(
+            "T2",
+            {"tensor(float16)",
+              "tensor(float)",
+              "tensor(double)",
+              "tensor(bfloat16)",
+              "tensor(uint8)",
+              "tensor(uint16)",
+              "tensor(uint32)",
+              "tensor(uint64)",
+              "tensor(int8)",
+              "tensor(int16)",
+              "tensor(int32)",
+              "tensor(int64)",
+              "tensor(bool)"},
+            "Constrain output types to all numeric tensors and bool tensors.")
+        .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+            if (ctx.getAttribute("dtype") != nullptr)
+              propagateElemTypeFromAttributeToOutput(ctx, "dtype", 0);
+            else
+              propagateElemTypeFromInputToOutput(ctx, 0, 0);
+            if (!hasNInputShapes(ctx, 1)) {
+              return;
+            }
+            propagateShapeFromInputToOutput(ctx, 0, 0);
+        })
+        .FunctionBody(FunctionBodyHelper::BuildNodes({
+            // nodes: {outputs, op, inputs, attributes}
+            {{"O1"}, 
+            "RandomUniform", 
+            {"input"},
+            {MakeAttribute("high", 1.0f), MakeAttribute("low", 0.f)}},
+            {{"O2"}, "Greater", {"01", "input"}},
```

Comment:
Change "01" to "O1"? Actually, alternative names like X1 or T1 may be better, since it is less likely to be confused with numbers like 01.

Reply:
Thanks for very useful comments!
Added the dtype attribute and cast output as well in FunctionBodyBuilde.